### PR TITLE
[TR] PIM-4248: Create Classification component and bundle to extract categories from CatalogBundle

### PR DESCRIPTION
**Pull request by willy-ahva**
| Q                    | A
| -------------------- | ---
| Bug fix?             | n
| New feature?         | no
| BC breaks?           | YES
| CI currently passes? | Running
| Tests pass?          | yes
| Scenarios pass?      | 
| Checkstyle issues?*  | no
| PMD issues?**        | no
| Changelog updated?   | yes
| Fixed tickets        | PIM-4248
| DB schema updated?   | no
| Migration script?    |
| Doc PR               |

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Entity/Category.php [hunk 1]
```diff
@@ -16,59 +17,12 @@
...
-
-    /**
-     * @var \Doctrine\Common\Collections\Collection
-     */
-    protected $children;
-
-    /**
-     * @var ArrayCollection
-     */
+    /** @var ArrayCollection */
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Entity/Category.php]: Please fill up the type of this ArrayCollection.
cf. http://docs.akeneo.com/master/contributing/standards.html#documentation

File: src/Pim/Bundle/CatalogBundle/Entity/Category.php [hunk 2]
```diff
@@ -76,206 +30,22 @@ class Category implements CategoryInterface
      */
     protected $locale;
 
-    /**
-     * @var ArrayCollection
-     */
+    /** @var ArrayCollection */
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Entity/Category.php]: Same

File: src/Pim/Bundle/CatalogBundle/Entity/Category.php [hunk 3]
```diff
@@ -76,206 +30,22 @@ class Category implements CategoryInterface
...
     protected $translations;
 
-    /**
-     * @var ArrayCollection
-     */
+    /** @var ArrayCollection */
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Entity/Category.php]: Same

File: src/Pim/Bundle/CatalogBundle/Model/CategoryInterface.php [hunk 4]
```diff
@@ -12,126 +13,13 @@
  * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
  * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
  */
-interface CategoryInterface extends TranslatableInterface, ReferableInterface, VersionableInterface
+interface CategoryInterface extends
+ BaseCategoryInterface,
+ TranslatableInterface,
+ ReferableInterface,
+ VersionableInterface
```
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Model/CategoryInterface.php]: It needs 4 spaces indentation

File: src/Pim/Bundle/ClassificationBundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ */
+class Configuration implements ConfigurationInterface
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function getConfigTreeBuilder()
+    {
+        $treeBuilder = new TreeBuilder();
+        $rootNode = $treeBuilder->root('pim_classification');
```
Comment (grena) [inline src/Pim/Bundle/ClassificationBundle/DependencyInjection/Configuration.php]: Delete this

File: src/Pim/Bundle/ClassificationBundle/DependencyInjection/PimClassificationExtension.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = new Configuration();
+        $config = $this->processConfiguration($configuration, $configs);
+
+        $loader = new Loader\YamlFileLoader($container, new FileLocator(__DIR__.'/../Resources/config'));
```
Comment (grena) [inline src/Pim/Bundle/ClassificationBundle/DependencyInjection/PimClassificationExtension.php]: You can leave the body of this method empty as it's obviously unused.

File: src/Pim/Bundle/CatalogBundle/Model/CategoryInterface.php (hunk 4 above)
Comment (willy-ahva) [inline src/Pim/Bundle/CatalogBundle/Model/CategoryInterface.php]: nope, phpcs told me only one

File: src/Pim/Component/Classification/Model/Category.php [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    /** @var int */
+    protected $right;
+
+    /** @var int */
+    protected $root;
+
+    /** @var CategoryInterface */
+    protected $parent;
+
+    /** @var \Doctrine\Common\Collections\Collection */
```
Comment (grena) [inline src/Pim/Component/Classification/Model/Category.php]: To put in the `use` statements.
Cf. http://docs.akeneo.com/master/contributing/standards.html#documentation

File: src/Pim/Component/Classification/Model/Category.php [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    {
+        return $this->children;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isRoot()
+    {
+        return ($this->getParent() === null);
```
Comment (grena) [inline src/Pim/Component/Classification/Model/Category.php]: Yoda

File: src/Pim/Component/Classification/Model/CategoryInterface.php [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+     * Predicate to know if this category has children
+     *
+     * @return bool
+     */
+    public function hasChildren();
+
+    /**
+     * Get children of this category
+     *
+     * @return \Doctrine\Common\Collections\Collection
```
Comment (grena) [inline src/Pim/Component/Classification/Model/CategoryInterface.php]: To put in the `use` statements

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+ * Category repository interface
+ *
+ * @author    Julien Janvier <jjanvier@akeneo.com>
+ * @copyright 2013 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface CategoryRepositoryInterface extends
+ TreeRepositoryInterface,
+ IdentifiableObjectRepositoryInterface,
+ ObjectRepository
```
Comment (grena) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: Needs 4 spaces

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+{
+    /**
+     * Get query builder for all existing category trees
+     *
+     * @return \Doctrine\ORM\QueryBuilder
```
Comment (grena) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: `use` statement

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+     */
+    public function getTreeFromParents(array $parentsIds);
+
+    /**
+     * Shortcut to get all children query builder
+     *
+     * @param CategoryInterface $category    the requested node
+     * @param bool              $includeNode true to include actual node in query result
+     *
+     * @return \Doctrine\ORM\QueryBuilder
```
Comment (grena) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: `use` statement

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    public function getCategoryIds(CategoryInterface $category, QueryBuilder $categoryQb = null);
+
+    /**
+     * Get children from a parent id
+     *
+     * @param int $parentId
+     *
+     * @return ArrayCollection
+     */
+    public function getChildrenByParentId($parentId);
```
Comment (grena) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: Its implementation returns an `array`.
Which one is wrong, implementation or interface ? Or raptors ?

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * Get children tree from a parent id.
+     * If the select node id is provided, the tree will be returned
+     * down to the node specified by select node id. Otherwise, the
+     * whole tree will be returned
+     *
+     * @param int $parentId
+     * @param int $selectNodeId
```
Comment (grena) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: Seems it's a `bool`

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+
+    /**
+     * Search Segment entities from an array of criterias.
+     * Search is done on a "%value%" LIKE expression.
+     * Criterias are joined with a AND operator
+     *
+     * @param int   $treeRootId Tree segment root id
+     * @param array $criterias  Criterias to apply
+     *
+     * @return ArrayCollection
```
Comment (grena) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: Same as above, its implementation returns an `array`.

File: src/Pim/Bundle/CatalogBundle/Model/CategoryInterface.php (hunk 4 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Model/CategoryInterface.php]: You used 4 here https://github.com/akeneo/pim-community-dev/pull/2760/files#diff-37662cce6ccad2bc286af5c0b8c29788R22

File: src/Pim/Bundle/ClassificationBundle/DependencyInjection/Configuration.php [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+use Symfony\Component\Config\Definition\ConfigurationInterface;
+
+/**
+ * Configuration
+ *
+ * @author    Willy Mesnage <willy.mesnage@akeneo.com>
+ * @copyright 2015 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class Configuration implements ConfigurationInterface
```
Comment (jjanvier) [inline src/Pim/Bundle/ClassificationBundle/DependencyInjection/Configuration.php]: remove this clas

File: src/Pim/Bundle/ClassificationBundle/DependencyInjection/PimClassificationExtension.php [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class PimClassificationExtension extends Extension
+{
+    /**
+     * {@inheritDoc}
+     */
+    public function load(array $configs, ContainerBuilder $container)
+    {
+        $configuration = new Configuration();
```
Comment (jjanvier) [inline src/Pim/Bundle/ClassificationBundle/DependencyInjection/PimClassificationExtension.php]: remove those 2 lines

File: src/Pim/Bundle/ClassificationBundle/DependencyInjection/PimClassificationExtension.php (hunk 6 above)
Comment (jjanvier) [inline src/Pim/Bundle/ClassificationBundle/DependencyInjection/PimClassificationExtension.php]: :+1:

File: src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php (hunk 13 above)
Comment (willy-ahva) [inline src/Pim/Component/Classification/Repository/CategoryRepositoryInterface.php]: Gedmo is wrong, it's an ArrayCollection

**Comment by grena** [comment]
GTM once travis green

**Pull request merged**